staging/src/k8s.io/cloud-provider/controllers/service/controller.go

```diff
@@ -765,26 +789,58 @@ func (s *Controller) nodeSyncInternal(ctx context.Context, workers int) {
-		return false
-	}
+// nodeSyncService syncs the nodes for one load balancer type service. The return value
+// indicates if we should retry. Hence, this functions returns false if we've updated
+// load balancers and finished doing it successfully. This function returns true if we
+// tried to update load balancers and failed, indicating to the caller that we should
+// try again.
+func (s *Controller) nodeSyncService(svc *v1.Service, newNodes []*v1.Node) bool {
+	newReadyNodes := filterWithPredicate(newNodes, s.getNodeConditionPredicate(includeOnlyReadyNodes))
 	klog.V(4).Infof("nodeSyncService started for service %s/%s", svc.Namespace, svc.Name)
-	hosts, err := listWithPredicate(s.nodeLister, s.getNodeConditionPredicate())
-	if err != nil {
-		runtime.HandleError(fmt.Errorf("failed to retrieve node list: %v", err))
-		return true
-	}
-
-	if err := s.lockedUpdateLoadBalancerHosts(svc, hosts); err != nil {
+	if err := s.lockedUpdateLoadBalancerHosts(svc, newReadyNodes); err != nil {
 		runtime.HandleError(fmt.Errorf("failed to update load balancer hosts for service %s/%s: %v", svc.Namespace, svc.Name, err))
 		return true
 	}
 	klog.V(4).Infof("nodeSyncService finished successfully for service %s/%s", svc.Namespace, svc.Name)
 	return false
 }
 
+// needSync computes if the service should be synced or not. It assumes the caller always
+// provides the current set of Ready and NotReady nodes as newNodes and the last synced set
+// of Ready and NotReady nodes as oldNodes. These are used to compute if ETP=local service
+// might need skipping or not.
+func (s *Controller) needsSync(svc *v1.Service, oldNodes, newNodes []*v1.Node) bool {
+	if svc == nil || !wantsLoadBalancer(svc) {
+		return false
+	}
+	// For services of ETP=local: we do not want to sync nodes transitioning from NotReady <-> Ready.
+	// The healthCheckNodePort probe executed by the LB will take care of dynamically ensuring traffic
+	// is sent to the right node(s). The "right node(s)" in this case are the node(s) hosting the endpoints.
+	// Either the node experiences a rapid change in transient state going from Ready -> NotReady -> Ready
+	// and there was no problem, or a change in "terminal state" going from Ready -> NotReady, upon which
+	// the endpoints eventually will be scheduled elsewhere. In any case: there's no need to re-configure
+	// the LB because the probe's result will give a more accurate representation of the state of things.
+	if svc.Spec.ExternalTrafficPolicy == v1.ServiceExternalTrafficPolicyTypeLocal {
+		// We have four cases to consider here:
+		//   a) Node ADD
+		//   b) Node DEL
+		//   c) Node UPDATE Ready -> NotReady
+		//   d) Node UPDATE NotReady -> Ready
+		// In case of (c) and (d) skip sync for ETP=Local Service
+		old := nodeNames(oldNodes)
+		new := nodeNames(newNodes)
+		if !old.Equal(new) {
+			// The set of nodes has changed - needs sync.
+			return true
+		}
+		// We are processing the same set of nodes, some might experience
```

Comment:
The filtering logic in this file is so spread around that I am having a really hard time convincing myself that we can't end up here when something else changed.

I'm looking for a way to make it clearer what's happening as the control moves around and we list and re-list and re-re-list with different criteria.

Let's describe the final end-state:

* We sync all nodes (except those labelled "exclude-from-lb") to all LB services.
* Stat of and changes to Node Unschedulable, Taints, readiness are all irrelevant

Right?  And the intermediate state is:

* We sync all nodes (except those labelled "exclude-from-lb") to ETP=Local LB services.
* We sync a subset of nodes (excluding Unschedulable, Taints, and unready) to other LB service
* Changes to Node Unschedulable, Taints, readiness are relvant to ETP!=Local

So can we set the code up to reflect this?  Like, just assume that every list of nodes includes ALL nodes down-filter only at the end, when we look at each Service?

Reply:
Let me see if I can summarize the logic as I've interpret it. Like you've said we react to node changes as:

> A Node was added
> A Node was deleted
> A Node was updated AND shouldSyncNode
>
> shouldSyncNode() is defined as any of:
>
> Node labels changed
> Node unschedulable changed
> Node ready changed

However `getNodeConditionPredicate` will perform a slightly different check, specifically looking at:

- Does the node has the "exclude label"
- Does it have the `ToBeDeletedTaint` key on its taints?
- Is it Ready or not 

Hence, as an example, we enqueue changes to `.Spec.Unschedulable`: but we don't consider it when actually syncing state. So even if it did change between two syncs, `getNodeConditionPredicate` doesn't check that attribute, so the set of nodes remain unchanged. What nodes get synced is the result of `getNodeConditionPredicate` only. 

I am reflecting on how big of an issue it is to not actually sync changes to `.Spec.Unschedulable`, given that we do sync changes to `.Spec.Taints` and specifically the: `ToBeDeletedTaint`. Maybe these go hand-in-hand? Doubt it...

> So can we set the code up to reflect this?

I am thinking about ways to make this better without changing too much.

Reply:
I'm OK with changing the code if it makes it simpler.  This code feel like it has evolved incorrectly - that we have different sets of conditions on different paths is a bit of a smell...